src/lifecycle/property.js

```diff
@@ -0,0 +1,149 @@
+import dashCase from '../util/dash-case';
+import data from '../util/data';
+import emit from '../api/emit';
+
+// TODO Decouple boolean attributes from the Boolean function.
+// TODO Split apart createNativePropertyDefinition function.
+
+function getLinkedAttribute (name, attr) {
+  return attr === true ? dashCase(name) : attr;
+}
+
+function createNativePropertyDefinition (name, opts) {
+  let prop = {
+    configurable: true,
+    enumerable: true
+  };
+
+  prop.created = function (elem, initialValue) {
+    let info = data(elem, `api/property/${name}`);
+    info.linkedAttribute = getLinkedAttribute(name, opts.attribute);
+    info.removeAttribute = elem.removeAttribute;
+    info.setAttribute = elem.setAttribute;
+    info.updatingProperty = false;
+
+    // TODO Refactor
+    if (info.linkedAttribute) {
+      if (!info.attributeMap) {
+        info.attributeMap = {};
+
+        elem.removeAttribute = function (attrName) {
+          info.updatingAttribute = true;
+          info.removeAttribute.call(this, attrName);
+          if (attrName in info.attributeMap) {
+            elem[info.attributeMap[attrName]] = undefined;
+          }
+          info.updatingAttribute = false;
+        };
+
+        elem.setAttribute = function (attrName, attrValue) {
+          info.updatingAttribute = true;
+          info.setAttribute.call(this, attrName, attrValue);
+          if (attrName in info.attributeMap) {
+            // Could also call getAttribute() but this does the same thing.
+            attrValue = String(attrValue);
+            elem[info.attributeMap[attrName]] = opts.deserialize ? opts.deserialize(attrValue) : attrValue;
```

Comment:
Perhaps `opts` should have a default value for `deserialize` and `serialize` that is just the identity function so that you don't have to check if it was specified and also so you don't forget call it anywhere (since you would just always call it)?